File: 백준/Silver/2178. 미로 탐색/미로 탐색.py
import sys
from collections import deque

n,m = map(int, input().split(" "))
matrix = [list(map(int, sys.stdin.readline().rstrip())) for i in range(n)]
visited = [[False] * m for _ in range(n)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def BFS(x, y):
    queue = deque()
    queue.append((x, y))

    # 큐가 빌 때까지 반복한다.
    while queue:
        x, y = queue.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            # 범위 밖일 경우 종료
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue

            # 0일 경우 종료
            if matrix[nx][ny] == 0:
                continue

            # 방문하지 않은 노드만 탐색
            if not visited[nx][ny]:
                visited[nx][ny] = True
                matrix[nx][ny] += matrix[x][y]
                queue.append((nx, ny))

        if x == n-1 and y == m-1:
            return matrix[x][y]

answer = BFS(0,0)
print(answer)

File: 백준/Silver/2178. 미로 탐색/미로 탐색.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class Main {
    public static int n;
    public static int m;
    public static int[][] matrix;
    public static int[][] visit;
    public static int[] dr = {-1, 1, 0, 0};
    public static int[] dc = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int[] input = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();
        n = input[0];
        m = input[1];

        // 입력
        matrix = new int[n + 1][m + 1];
        visit = new int[n + 1][m + 1];
        for (int i = 1; i < n + 1; i++) {
            String str = br.readLine();
            for (int j = 1; j < m + 1; j++) {
                int x = str.charAt(j - 1) - '0';
                matrix[i][j] = x;
            }
        }

        bfs(1, 1);
    }

    public static void bfs(int r, int c) {
        visit[r][c] = 1;
        Queue<D_2178> q = new LinkedList<>();
        q.add(new D_2178(r, c));

        while (!q.isEmpty()) {
            D_2178 current = q.poll();
            for (int k = 0; k < 4; k++) {
                int nr = current.row + dr[k];
                int nc = current.col + dc[k];

                if (0 > nr || nr > n || 0 > nc || nc > m) continue;

                if (matrix[nr][nc] == 1 && visit[nr][nc] == 0) {
                    q.add(new D_2178(nr, nc));
                    visit[nr][nc] = visit[current.row][current.col] + 1;
                }
            }
        }

        System.out.println(visit[n][m]);
    }
}

class D_2178 {
    int row;
    int col;

    public D_2178(int row, int col) {
        this.row = row;
        this.col = col;
    }
}